sdk/cosmos/azure-cosmos-spark_3/src/main/scala/com/azure/cosmos/spark/TransactionalBatchWriter.scala

```diff
@@ -0,0 +1,509 @@
+ *   If not provided, the default operation is "upsert".
+ * 
+ * Output DataFrame schema:
+ * - id: String
+ * - partitionKey: String
+ * - operationType: String
+ * - statusCode: Int
+ * - isSuccessStatusCode: Boolean
+ * - resultDocument: String (optional - for read operations)
+ * - errorMessage: String (optional)
+ */
+private[spark] class TransactionalBatchWriter(
+  val outputSchema: StructType,
+  val userConfig: Map[String, String]
+) extends BasicLoggingTrait with Serializable {
+
+  val effectiveUserConfig: Map[String, String] = CosmosConfig.getEffectiveConfig(
+    databaseName = None,
+    containerName = None,
+    userConfig)
+
+  val clientConfig: CosmosAccountConfig = CosmosAccountConfig.parseCosmosAccountConfig(effectiveUserConfig)
+  val readConfig: CosmosReadConfig = CosmosReadConfig.parseCosmosReadConfig(effectiveUserConfig)
+  val cosmosContainerConfig: CosmosContainerConfig =
+    CosmosContainerConfig.parseCosmosContainerConfig(effectiveUserConfig)
+
+  val tableName: String = s"com.azure.cosmos.spark.items.${clientConfig.accountName}." +
+    s"${cosmosContainerConfig.database}.${cosmosContainerConfig.container}"
+
+  private lazy val sparkSession = {
+    assertOnSparkDriver()
+    SparkSession.active
+  }
+
+  val sparkEnvironmentInfo: String = CosmosClientConfiguration.getSparkEnvironmentInfo(Some(sparkSession))
+  logTrace(s"Instantiated ${this.getClass.getSimpleName} for $tableName")
+
+  private[spark] def initializeAndBroadcastCosmosClientStatesForContainer(): (Broadcast[CosmosClientMetadataCachesSnapshots], Broadcast[String]) = {
+    val calledFrom = s"TransactionalBatchWriter($tableName).initializeAndBroadcastCosmosClientStateForContainer"
+    Loan(
+      List[Option[CosmosClientCacheItem]](
+        Some(
+          CosmosClientCache(
+            CosmosClientConfiguration(
+              effectiveUserConfig,
+              readConsistencyStrategy = readConfig.readConsistencyStrategy,
+              sparkEnvironmentInfo),
+            None,
+            calledFrom)),
+        ThroughputControlHelper.getThroughputControlClientCacheItem(
+          effectiveUserConfig,
+          calledFrom,
+          None,
+          sparkEnvironmentInfo)
+      ))
+      .to(clientCacheItems => {
+        val container =
+          ThroughputControlHelper.getContainer(
+            effectiveUserConfig,
+            cosmosContainerConfig,
+            clientCacheItems(0).get,
+            clientCacheItems(1))
+        
+        // Read container properties to get partition key definition
+        val containerProperties: CosmosContainerProperties = 
+          SparkBridgeInternal.getContainerPropertiesFromCollectionCache(container)
+        val partitionKeyDefinition: PartitionKeyDefinition = 
+          containerProperties.getPartitionKeyDefinition()
+        
+        // Warm up the client by attempting a read
```

Comment:
why needed? Just results in noise 404 that some cx gest confused about?

Reply:
I think I saw what looked like the same pattern in CosmosReadManyReader and thought it was being used to "warm up" avoid cold-start penalty. Can remove.

Reply:
Removed.